File: leetcode/Matrix/0036_Valid_Sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
    	vector<bool> checked(10, false);
        // check every row
        for (int i = 0; i < 9; i++) {
        	for (int j = 0; j < 9; j++) {
        		if (board[i][j] == '.')
        			continue;
        		else if (checked[board[i][j] - '0'])
        			return false;
        		else
        			checked[board[i][j] - '0'] = true;
        	}
        	checked = vector<bool> (10, false);
        }

        // check every column
        for (int i = 0; i < 9; i++) {
        	for (int j = 0; j < 9; j++) {
        		if (board[j][i] == '.')
        			continue;
        		else if (checked[board[j][i] - '0'])
        			return false;
        		else
        			checked[board[j][i] - '0'] = true;
        	}
        	checked = vector<bool> (10, false);
        }

        // check every sub-boxes
        for (int i = 0; i < 3; i++) {
        	for (int j = 0; j < 3; j++) {
        		for (int k = 0; k < 3; k++) {
        			for (int l = 0; l < 3; l++) {
        				if (board[i * 3 + k][j * 3 + l] == '.')
        					continue;
        				else if (checked[board[i * 3 + k][j * 3 + l] - '0'])
		        			return false;
		        		else
		        			checked[board[i * 3 + k][j * 3 + l] - '0'] = true;
        			}
        		}
        		checked = vector<bool> (10, false);
        	}
        }
        return true;
    }
};

File: leetcode/Matrix/0036_Valid_Sudoku.py
class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        # check rows
        for i in range(9):
            num_set = set()
            for j in range(9):
                if board[i][j] != '.':
                    if board[i][j] in num_set:
                        return False
                    else:
                        num_set.add(board[i][j])

        # check columns
        for j in range(9):
            num_set = set()
            for i in range(9):
                if board[i][j] != '.':
                    if board[i][j] in num_set:
                        return False
                    else:
                        num_set.add(board[i][j])

        # check 3 * 3 sub-box
        for i in range(3):
            for j in range(3):
                num_set = set()
                sub_board = [
                    board[i * 3][j * 3], board[i * 3 + 1][j * 3], board[i * 3 + 2][j * 3],
                    board[i * 3][j * 3 + 1], board[i * 3 + 1][j * 3 + 1], board[i * 3 + 2][j * 3 + 1],
                    board[i * 3][j * 3 + 2], board[i * 3 + 1][j * 3 + 2], board[i * 3 + 2][j * 3 + 2],
                ]
                for char in sub_board:
                    if char != '.':
                        if char in num_set:
                            return False
                        else:
                            num_set.add(char)

        return True